controllers/dashboard_controller.go

```diff
@@ -0,0 +1,329 @@
+		}
+	}
+
+	return ctrl.Result{RequeueAfter: RequeueDelaySuccess}, nil
+}
+
+func (r *GrafanaDashboardReconciler) reconcileDashboardInInstance(ctx context.Context, grafana *v1beta1.Grafana, cr *v1beta1.GrafanaDashboard) error {
+	if strings.TrimSpace(cr.Spec.Json) == "" {
+		return nil
+	}
+
+	var dashboardFromJson map[string]interface{}
+
+	err := json.Unmarshal([]byte(cr.Spec.Json), &dashboardFromJson)
+	if err != nil {
+		return err
+	}
+
+	grafanaClient, err := client2.NewGrafanaClient(ctx, r.Client, grafana)
+	if err != nil {
+		r.Log.Error(err, "error establishing new api client for instance", "grafana", grafana.Name, "namespace", grafana.Namespace, "dashboard", cr.Name)
+		return err
+	}
+
+	dashboards, err := grafanaClient.Dashboards()
+	if err != nil {
+		r.Log.Error(err, "could not get dashboards")
+		return err
+	}
+	logger.Log.Info("found dashboards", "dashboards", dashboards)
+
+	status := cr.Status.DeepCopy()
+
+	if status.UIDforInstance == nil {
+		status.UIDforInstance = make(map[string]string)
+	}
+
+	getDashUIDFromStatus := func(namespacedName string) string {
+		if status != nil {
+			for grafanaInstance, dashboardUID := range status.UIDforInstance {
+				if grafanaInstance == namespacedName {
+					return dashboardUID
+				}
+			}
+		}
+		return ""
+	}
+
+	grafanaNamespacedName := fmt.Sprintf("%s%s", grafana.Namespace, grafana.Name)
+	dashUIDfromStatus := getDashUIDFromStatus(grafanaNamespacedName)
+
+	if cr.DeletionTimestamp != nil {
+		r.Log.Info("found a deletion timestamp for dashboard", "dashboard", cr.ObjectMeta.Name, "instance", grafanaNamespacedName)
+		if err := grafanaClient.DeleteDashboardByUID(dashUIDfromStatus); strings.Contains(fmt.Sprintf("%v", err), "404") {
+			r.Log.Info("the dashboard couldn't be found on the instance, nothing to do, removing finalizer")
+			controllerutil.RemoveFinalizer(cr, dashboardFinalizer)
+			err = r.Client.Update(ctx, cr)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+		if err != nil {
+			r.Log.Error(err, "could not delete dashboard from instance", "instance", grafanaNamespacedName)
+			return err
+		} else {
+			delete(status.UIDforInstance, grafanaNamespacedName)
+
+			if err := r.reconcileDashboardStatus(cr, status); err != nil {
+				r.Log.Error(err, "error while updating dashboard status")
+				return err
+			}
+			if status == nil {
+				controllerutil.RemoveFinalizer(cr, dashboardFinalizer)
+			}
+		}
+		return nil
+	}
+
+	// dashboard is not known try to create it
+	if dashUIDfromStatus == "" {
+		resp, err := grafanaClient.NewDashboard(grapi.Dashboard{
+			Meta: grapi.DashboardMeta{
+				IsStarred: false,
+				Slug:      cr.ObjectMeta.Name,
+				//Folder:    ,
+				//URL:       "",
+			},
+			Model: dashboardFromJson,
+			//Folder:    0,
+			Overwrite: false,
+			Message:   "",
+		})
+
+		if err != nil {
+			if !strings.Contains(fmt.Sprintf("%s", err), statusCodeDashboardExistsInFolder) {
```

Comment:
Since `error` is an interface with `Error()` method that returns its text representation ([source code](https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/builtin/builtin.go;drc=f524a2baa4aca9e603fbf3891951b53b497dc0a1;l=268)), you might prefer something like:
```go
if !strings.Contains(err.Error(), statusCodeDashboardExistsInFolder)
```
instead of:
```go
if !strings.Contains(fmt.Sprintf("%s", err), statusCodeDashboardExistsInFolder)
```

Also, since there will be lots and lots of places where we'll need to inspect particular error codes (e.g. 2 places just in your PR), I'd suggest to turn it all into a little helper. Plus, to make sure that we always inspect status code, not body contents of the error (maybe I'm too careful with this, and numbers never appear there), I'll make the helper close to grafana api client error format, which is:
```go
if resp.StatusCode >= 400 {
    return fmt.Errorf("status: %d, body: %v", resp.StatusCode, string(bodyContents))
}
```
([source](https://github.com/grafana/grafana-api-golang-client/blob/9c38b6f437bb9688f655ace9e66cc8b8f91ce0f4/client.go#L119))

An example of how it could be implemented (if you like the idea ofc):
```go
func errorStatusCodeIs(err error, code string) bool {
	prefix := fmt.Sprintf("status: %s,", code)
	return strings.HasPrefix(err.Error(), prefix)
}
```

Reply:
+1 on checking the status code instead of the string representation